onnxruntime/python/tools/quantization/calibrate.py

```diff
@@ -0,0 +1,232 @@
+    '''
+    Load tensor data from pb files.
+    :param data_file_name: path to the pb file
+    :param dataset_size: number of image-data in the pb file for data size check
+    :param height: image height for data size check
+    :param width: image width for data size check
+    :return input data for the model
+    '''
+    tensor = onnx.TensorProto()
+    inputs = np.empty(0)
+    with open(data_file_name, 'rb') as fin:
+        tensor.ParseFromString(fin.read())
+        inputs = numpy_helper.to_array(tensor)
+        try:
+            shape = inputs.shape
+            inputs = inputs.reshape(dataset_size, samples, channels, height, width)
+        except:
+            sys.exit("Input .pb file contains incorrect input size. \nThe required size is: (%s). The real size is: (%s)"
+                        %((dataset_size, samples, channels,height, width), shape))
+
+    return inputs
+
+def main():
+    # Parsing command-line arguments
+    parser = argparse.ArgumentParser(description='parsing model and test data set paths')
+    parser.add_argument('--model_path', required=True)
+    parser.add_argument('--dataset_path', required=True)
+    parser.add_argument('--calib_mode', default='naive')
+    parser.add_argument('--dataset_size', type=int, default=30)
+    parser.add_argument('--data_preprocess', type=str, default='')
+    args = parser.parse_args()
+    model_path = args.model_path
+    images_folder = args.dataset_path
+    calib_mode = args.calib_mode
+    size_limit = args.dataset_size
+
+    # Generating augmented ONNX model
+    augmented_model_path = 'augmented_model.onnx'
+    model = onnx.load(model_path)
+    augmented_model = augment_graph(model)
+    onnx.save(augmented_model, augmented_model_path)
+
+    # Conducting inference
+    session = onnxruntime.InferenceSession(augmented_model_path, None)
+    (samples, channels, height, width) = session.get_inputs()[0].shape
+
+    # Generating inputs for quantization
+    if args.data_preprocess:
+        inputs = load_batch(images_folder, height, width, size_limit, args.data_preprocess)
+    else:
+        inputs = load_pb_file(images_folder, args.dataset_size, samples, channels, height, width)
+    print(inputs.shape)
+    dict_for_quantization = get_intermediate_outputs(model_path, session, inputs, calib_mode)
+    quantization_params_dict = calculate_quantization_params(model, quantization_thresholds=dict_for_quantization)
+    calibrated_quantized_model = quantize(onnx.load(model_path), quantization_mode=QuantizationMode.QLinearOps, quantization_params=quantization_params_dict)
+    onnx.save(calibrated_quantized_model, 'calibrated_quantized_model.onnx')
+
+    print("Calibrated, quantized model saved.")
+
+# Candidate nodes for quantization. Calibration will be done for these nodes only
+# When more nodes are extended to support quantization, add them to this list
+quantization_candidates = ['Conv', 'MatMul']
```

Comment:
does it need to be global?

Reply:
variable defined outside a function is global by default. global keyword will be required if we were to create this inside a function or modify this inside the function

Reply:
i meant, it's only used in a single function, could it be made local to that function?
globals are kind of discouraged.